qemu/tests/rng_stress.py

```diff
@@ -0,0 +1,78 @@
+import logging
+import aexpect
+
+from virttest import error_context
+from virttest import utils_test
+from virttest.qemu_devices import qdevices
+from avocado.core import exceptions
+from autotest.client.shared import utils
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    1) Boot up vm with one or more rng devices
+    2) Read from /dev/random in host (optional)
+    3) Run random read in guest
+    4) Switch rng device if there is more than on rng devices
+    5) Run random read in guest
+    6) Read from /dev/random in host (optional)
+    7) Clean random read in host (optional)
+    """
+
+    def get_rng_list(vm):
+        """
+        Get attached rng devices from device dictionary
+        """
+        rng_list = []
+        for device in vm.devices:
+            if isinstance(device, qdevices.QDevice):
+                if device.get_param("driver") == "virtio-rng-pci":
+                    rng_list.append(device)
+        return rng_list
+
+    def get_available_rng(session):
+        """
+        Get available rng devices from /sys/devices
+        """
+        verify_cmd = params.get["driver_available_cmd"]
+        try:
+            rng_devices = session.cmd_output_status(verify_cmd)
+        except aexpect.ShellTimeoutError:
+            err = "cat cmd timeout, pls check if it's a product bug"
+            raise exceptions.TestFail(err)
+        rng_devices = rng_devices.split()
+        return rng_devices
+
+    login_timeout = int(params.get("login_timeout", 360))
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=login_timeout)
+    sub_test = params.get("sub_test")
+
+    if params.get("pre_cmd"):
+        error_context.context("Fetch data from host", logging.info)
+        utils.system(params.get("pre_cmd"))
+
+    error_context.context("Read rng device in guest", logging.info)
+    utils_test.run_virt_sub_test(test, params, env, sub_test)
+
+    if params.get("os_type") == "linux":
+        error_context.context("Query virtio rng device in guest",
+                              logging.info)
+        rng_devices = get_available_rng(session)
+        rng_attached = get_rng_list(vm)
+        if len(rng_devices) != len(rng_attached):
+            raise exceptions.TestFail("The devices get from rng_arriable"
+                                      " don't match the rng devices attached")
+
+        for rng_device in rng_devices:
+            error_context.context("Change virtio rng device to %s" %
+                                  rng_device, logging.info)
+            session.cmd_status(params.get("switch_rng_cmd") % rng_device)
+            error_context.context("Read from %s in guest" % rng_device,
+                                  logging.info)
+            utils_test.run_virt_sub_test(test, params, env, sub_test)
+
+    if params.get("post_cmd"):
+        utils.system(params.get("post_cmd"))
```

Comment:
for case one_device, the read size of dd is too small to check dd process.
05:34:04 ERROR| Traceback (most recent call last):
05:34:04 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-41.0-py2.7.egg/avocado_vt/test.py", line 204, in runTest
05:34:04 ERROR|     raise exceptions.TestError(details)
05:34:04 ERROR| TestError: Command <kill -9 `ps -a -o ppid,comm | grep dd | awk '{print $1}'`> failed, rc=1, Command returned non-zero exit status    [context: Read rng device in guest]
05:34:04 ERROR| * Command:
05:34:04 ERROR|     kill -9 `ps -a -o ppid,comm | grep dd | awk '{print $1}'`
05:34:04 ERROR| Exit status: 1